mixxxdj/mixxx
# Add support for Serato Cue Export

Whohoo, it works:

![Mixxx Cues in Serato](https://user-images.githubusercontent.com/1834516/93261999-cc726380-f7a3-11ea-841b-789b28a1f082.png)

This PR lacks the "glue" code in the `Track::exportMetadata` method because that needs more thought. We need to find a poper way to get the signal info (to calculate cue positions in frames and find the timing offset) before we can merge this. For testing, I implemented an extremely hacky version (see the diff below).

**We should implement that code in a separate PR to keep the diff small - this one is ready to review/merge IMHO.**

## Testing instructions

First apply the patch below. Then you can run mixxx, load a track to a deck, set some cues, select "Export metadata to file tags", eject the track and close mixxx. The cues should then show up in Serato. 

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 45cd5ab1ca..3e1e3f0ea7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -915,7 +915,7 @@ add_library(mixxx-lib STATIC EXCLUDE_FROM_ALL
 )
 set_target_properties(mixxx-lib PROPERTIES AUTOMOC ON AUTOUIC ON)
 target_include_directories(mixxx-lib PUBLIC src "${CMAKE_CURRENT_BINARY_DIR}/src")
-target_compile_definitions(mixxx-lib PRIVATE SETTINGS_FILE="mixxx.cfg")
+target_compile_definitions(mixxx-lib PRIVATE SETTINGS_FILE="mixxx.cfg" __EXPORT_SERATO_MARKERS__)
 if(UNIX AND NOT APPLE)
   target_compile_definitions(mixxx-lib PRIVATE SETTINGS_PATH=".mixxx/")
 endif()
diff --git a/src/track/track.cpp b/src/track/track.cpp
index b18487337b..495ac760bc 100644
--- a/src/track/track.cpp
+++ b/src/track/track.cpp
@@ -1341,6 +1341,20 @@ ExportTrackMetadataResult Track::exportMetadata(
         mixxx::SeratoTags* seratoTags = m_record.refMetadata().refTrackInfo().ptrSeratoTags();
         seratoTags->setTrackColor(getColor());
         seratoTags->setBpmLocked(isBpmLocked());
+
+        DEBUG_ASSERT(m_streamInfo);
+        if (m_streamInfo) {
+            mixxx::audio::SampleRate sampleRate = m_streamInfo->getSignalInfo().getSampleRate();
+
+            QList<mixxx::CueInfo> cueInfos;
+            for (const CuePointer& pCue : m_cuePoints) {
+                cueInfos.append(pCue->getCueInfo(sampleRate));
+            }
+
+            double timingOffset = mixxx::SeratoTags::guessTimingOffsetMillis(
+                    getLocation(), m_streamInfo->getSignalInfo());
+            seratoTags->setCueInfos(cueInfos, timingOffset);
+        }
     }
 #endif
```
